client/material/tabs.tsx

```diff
@@ -174,5 +177,52 @@ export function Tabs<T>({ children, activeTab, onChange, className }: TabsProps<
     return tabElems
   }, [activeTab, children, onChange])
 
+  useKeyListener({
+    onKeyDown: (event: KeyboardEvent) => {
+      if ((event.keyCode === PAGEUP || event.keyCode === PAGEDOWN) && event.ctrlKey === true) {
+        const enabledChildren: React.ReactElement[] = []
+        React.Children.forEach(children, (child, i) => {
+          if (child && !(child.props as TabItemProps<T>).disabled) {
+            enabledChildren.push(child)
+          }
+        })
+
+        if (enabledChildren.length < 2) {
+          return false
+        }
+
+        let previousOrNextTab: T | undefined
+        React.Children.forEach(enabledChildren, (child, i) => {
+          let previousOrNextChild
+          if (event.keyCode === PAGEUP) {
+            previousOrNextChild =
+              i === enabledChildren.length - 1 ? enabledChildren[0] : enabledChildren[i + 1]
+          } else {
+            previousOrNextChild =
+              i === 0 ? enabledChildren[enabledChildren.length - 1] : enabledChildren[i - 1]
+          }
+
+          if (
+            child &&
+            previousOrNextChild &&
+            (previousOrNextChild.props as TabItemProps<T>).value === activeTab
```

Comment:
I can't really understand what this check is meant to do? Can you explain what you're trying to accomplish here?

To me this reads as "if the tab we're going to is already the active one, send an onChange event for it" which seems... backwards

Reply:
It is very likely that I made this code more convoluted than it needs to be, but my thinking was that in order to find the previous/next tab to switch to, I have to iterate through all the (enabled) tabs and check if the active tab is the one in front/behind of the one I'm looking for. This also explains why the logic above seems backwards. For example, to find the _previous_ tab, I have to check if the _next_ tab is the active one.

Reply:
We should step back and rethink this whole thing then, I don't think this is really maintainable as-is and is very likely to be a source of bugs going forward

Reply:
So the goal here is, "call onChange with the value of the tab that should be newly selected". So first thing's first, this function can exit early if onChange isn't set, there's nothing useful we can do in that case.

Then, when you loop through the children to make an `enabledChildren` list, *always* include the currently selected tab whether it is enabled or not. Store the index of that tab when you find it.

Next, we made an array of all enabled tabs above. There's no need to loop through things again, we have all that we need in that array:

```ts
if (event.keyCode === PAGEUP) {
  const previousIndex = (selectedIndex - 1 + enabledChildren.length) % enabledChildren.length
  const tab = enabledChildren[previousIndex]
  onChange(tab.props.value)
} else if (event.keycode === PAGEDOWN) {
  const nextIndex = selectedIndex + 1 % enabledChildren.length
  const tab = enabledChildren[nextIndex]
  onChange(tab.props.value)
}
```

Much simpler